compaction/compaction_manager.cc

```diff
@@ -1127,12 +1127,21 @@ future<> compaction_manager::drain() {
 }
 
 future<> compaction_manager::stop() {
-    do_stop();
-    if (auto cm = std::exchange(_task_manager_module, nullptr)) {
-        co_await cm->stop();
-    }
-    if (_state != state::none) {
-        co_return co_await std::move(*_stop_future);
+    try {
+        do_stop();
```

Comment:
do_stop() should prevent throwing on its own.

Reply:
it does.  I'm afraid the error came from the task manager stop() call, not from do_stop

Reply:
Then keep do_stop() outside the try/catch, and fix task_manager::stop.

Reply:
The error comes from `co_return co_await std::move(*_stop_future);`:

`do_stop` sets `_stop_future` to `really_do_stop`
`do_really_stop` co_awaits `stop_ongoing_compactions`

https://github.com/scylladb/scylladb/blob/88e2a7ea192b5963ee03f5ce4975d355e28c32c4/compaction/compaction_manager.cc#L1154

And `stop_ongoing_compactions` has noexcept trick:
https://github.com/scylladb/scylladb/blob/88e2a7ea192b5963ee03f5ce4975d355e28c32c4/compaction/compaction_manager.cc#L1112-L1114

Reply:
Slightly off topic, but isn't a noexcept function returning an exception future somewhat breaking its contract (at least w.r. to a co_await caller)?

Reply:
> Slightly off topic, but isn't a noexcept function returning an exception future somewhat breaking its contract (at least w.r. to a co_await caller)?

Technically no, it _returns_ and not _throws_.
The contract here is that the `future` may contain many things, one of them is an exception.
So in this sense this is also not a contract violation.

Reply:
Agreed with @regevran.  There's no equivalent for `noexcept` for an async function that may return no errors.

Reply:
exceptions are evil

Reply:
but evil isn't an exception